src/data_source.rs

```diff
@@ -769,4 +266,70 @@ mod test {
             );
         }
     }
+
+    #[async_std::test]
+    pub async fn test_range<D: TestableDataSource>() {
+        setup_test();
+
+        let mut network = MockNetwork::<D>::init().await;
+        let qd = network.query_data();
+        network.start().await;
+
+        // Wait for there to be at least 3 blocks, then lock the state so the block height can't
+        // change during the test.
+        let (qd, block_height) = loop {
+            let qd = qd.read().await;
+            let block_height = qd.block_height().await.unwrap();
+            if block_height >= 3 {
+                break (qd, block_height as u64);
+            }
+        };
+
+        // Query for a variety of ranges testing all cases of included, excluded, and unbounded
+        // starting and ending bounds
+        do_range_test(&*qd, 1..=2, 1..3).await; // (inclusive, inclusive)
+        do_range_test(&*qd, 1..3, 1..3).await; // (inclusive, exclusive)
+        do_range_test(&*qd, 1.., 1..block_height).await; // (inclusive, unbounded)
+        do_range_test(&*qd, ..=2, 0..3).await; // (unbounded, inclusive)
+        do_range_test(&*qd, ..3, 0..3).await; // (unbounded, exclusive)
+        do_range_test(&*qd, .., 0..block_height).await; // (unbounded, unbounded)
+        do_range_test(&*qd, ExRange(0..=2), 1..3).await; // (exclusive, inclusive)
+        do_range_test(&*qd, ExRange(0..3), 1..3).await; // (exclusive, exclusive)
+        do_range_test(&*qd, ExRange(0..), 1..block_height).await; // (exclusive, unbounded)
```

Comment:
Aren't these three tests equivalent to the first three after the ExRange wrapping?

Reply:
In other words, what exactly are we testing here?

Reply:
The ranges being queried are the same but the ranges are being described in a different way. The first three are testing how we handle an inclusive lower bound and these three are testing how we handle an exclusive lower bound.